src/client/keycode.cpp

```diff
@@ -221,122 +223,172 @@ static const struct table_key table[] = {
 
-	std::ostringstream os;
-	os << "<Char " << hex_encode((char*) &Char, sizeof(wchar_t)) << ">";
-	throw UnknownKeycode(os.str().c_str());
+	auto wname = utf8_to_wide(name);
+	if (wname.empty())
+		return invalid_key;
+	return lookup_keychar(wname[0]);
 }
 
-KeyPress::KeyPress(const char *name)
+static const table_key &lookup_scancode(const u32 scancode)
 {
-	if (strlen(name) == 0) {
-		Key = irr::KEY_KEY_CODES_COUNT;
-		Char = L'\0';
-		m_name = "";
-		return;
+	auto device = RenderingEngine::get_raw_device();
+	if (!device) {
+		warningstream << "IrrlichtDevice not available during scancode lookup." <<
+				" The resulting value is invalid." << std::endl;
+		return invalid_key;
 	}
 
-	if (strlen(name) <= 4) {
-		// Lookup by resulting character
-		int chars_read = mbtowc(&Char, name, 1);
-		FATAL_ERROR_IF(chars_read != 1, "Unexpected multibyte character");
-		try {
-			auto &k = lookup_keychar(Char);
-			m_name = k.Name;
-			Key = k.Key;
-			return;
-		} catch (UnknownKeycode &e) {};
+	auto key = device->getKeyFromScancode(scancode);
+	return std::holds_alternative<EKEY_CODE>(key) ?
+		lookup_keykey(std::get<irr::EKEY_CODE>(key)) :
+		lookup_keychar(std::get<wchar_t>(key));
+}
+
+static const table_key &lookup_scancode(const std::variant<u32, irr::EKEY_CODE> &scancode)
+{
+	return std::holds_alternative<irr::EKEY_CODE>(scancode) ?
+		lookup_keykey(std::get<irr::EKEY_CODE>(scancode)) :
+		lookup_scancode(std::get<u32>(scancode));
+}
+
+void KeyPress::loadFromKey(irr::EKEY_CODE keycode, wchar_t keychar)
+{
+	if (auto device = RenderingEngine::get_raw_device()) {
+		scancode = device->getScancodeFromKey(Keycode(keycode, keychar));
 	} else {
-		// Lookup by name
-		m_name = name;
-		try {
-			auto &k = lookup_keyname(name);
-			Key = k.Key;
-			Char = k.Char;
-			return;
-		} catch (UnknownKeycode &e) {};
+		warningstream << "IrrlichtDevice not available when creating KeyPress." <<
+				" The resulting value may be incorrect or unusable." << std::endl;
+		if (Keycode::isValid(keycode))
+			scancode = keycode;
+		else
+			scancode = (u32) keychar;
 	}
+}
 
-	// It's not a known key, complain and try to do something
-	Key = irr::KEY_KEY_CODES_COUNT;
-	int chars_read = mbtowc(&Char, name, 1);
-	FATAL_ERROR_IF(chars_read != 1, "Unexpected multibyte character");
-	m_name = "";
-	warningstream << "KeyPress: Unknown key '" << name
-		<< "', falling back to first char." << std::endl;
+KeyPress::KeyPress(const std::string &name)
+{
+	if (loadFromScancode(name))
+		return;
+	const auto &key = lookup_keyname(name);
+	loadFromKey(key.Key, key.Char);
 }
 
-KeyPress::KeyPress(const irr::SEvent::SKeyInput &in, bool prefer_character)
+KeyPress::KeyPress(const irr::SEvent::SKeyInput &in)
 {
-	if (prefer_character)
-		Key = irr::KEY_KEY_CODES_COUNT;
+#if USE_SDL2
+	if (in.SystemKeyCode)
+		scancode.emplace<u32>(in.SystemKeyCode);
 	else
-		Key = in.Key;
-	Char = in.Char;
+		scancode.emplace<irr::EKEY_CODE>(in.Key);
+#else
+	loadFromKey(in.Key, in.Char);
+#endif
```

Comment:
Cannot add a code suggestion... thanks GitHub.
```C++
	if (USE_SDL2) {
		if (in.SystemKeyCode)
			scancode.emplace<u32>(in.SystemKeyCode);
		else
			scancode.emplace<irr::EKEY_CODE>(in.Key);
	} else {
		loadFromKey(in.Key, in.Char);
	}
```
Same idea as in `formatScancode` below.

Reply:
https://github.com/luanti-org/luanti/pull/14964/commits/c54a626fe3142c71c770563084471344adc3e6df